### PR TITLE
44414 data frame attributes not propagating with astype

### LINKED ISSUE #44414: BUG: DataFrame attributes are lost after specifying dtype for select columns
### 

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [ ] I have confirmed this bug exists on the master branch of pandas.

### Reproducible Example

```python
# I add attributes to a DataFrame
df = pd.DataFrame(dict(column1=[1,2,3,4], column2=[1,2,3,4]))
df.attrs['name'] = "hello"
print(df)
print(df.attrs)

OUT:
   column1  column2
0        1        1
1        2        2
2        3        3
3        4        4
{'name': 'hello'}

# When I change the dtype, the attribute is still there (as expected)
df1 = df.astype(float)
print(df1)
print(df1.attrs)

OUT:
   column1  column2
0      1.0      1.0
1      2.0      2.0
2      3.0      3.0
3      4.0      4.0
{'name': 'hello'}

# But when I specify the dtype for specific columns, the attributes are lost
df2 = df.astype({'column1':float})
print(df2)
print(df2.attrs)

OUT:
column1  column2
0      1.0        1
1      2.0        2
2      3.0        3
3      4.0        4
{}
```

### Issue Description

I assigned attributes to a DataFrame. When I change the DataFrame dtype with `df.astype(float)`, the DataFrame attribute are retained, as expected. But when I change the dtype for specific columns with `df.astype({'column1':float})`, the attributes are lost.

### Expected Behavior

I expected the DataFrame to retrain the attributes after specifying the dtype for specific columns as it does when I specify the dtype for the full DataFrame.

### Installed Versions

<details>

INSTALLED VERSIONS
------------------
commit           : 945c9ed766a61c7d2c0a7cbb251b6edebf9cb7d5
python           : 3.9.7.final.0
python-bits      : 64
OS               : Linux
OS-release       : 5.3.18-24.78-default
Version          : #1 SMP Mon Aug 2 21:00:23 UTC 2021 (a13100d)
machine          : x86_64
processor        : x86_64
byteorder        : little
LC_ALL           : None
LANG             : en_US.UTF-8
LOCALE           : en_US.UTF-8

pandas           : 1.3.4
numpy            : 1.21.3
pytz             : 2021.3
dateutil         : 2.8.2
pip              : 21.3.1
setuptools       : 58.2.0
Cython           : None
pytest           : 6.2.5
hypothesis       : None
sphinx           : 4.0.3
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : None
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : 2.11.0
IPython          : 7.28.0
pandas_datareader: None
bs4              : 4.10.0
bottleneck       : None
fsspec           : 2021.10.1
fastparquet      : None
gcsfs            : None
matplotlib       : 3.4.3
numexpr          : None
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : None
pyxlsb           : None
s3fs             : 2021.10.1
scipy            : 1.7.1
sqlalchemy       : 1.4.26
tables           : None
tabulate         : None
xarray           : 0.19.0
xlrd             : None
xlwt             : None
numba            : None

</details>

### PR DESCRIPTION
**Pull request by skwirskj**
- [x] closes #44414 
- [x] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit) for how to run them
- [x] whatsnew entry

### CODE REVIEW

File: doc/source/whatsnew/v1.4.0.rst [hunk 1]
```diff
@@ -636,7 +636,7 @@ Conversion
 - Bug in :class:`IntegerDtype` not allowing coercion from string dtype (:issue:`25472`)
 - Bug in :func:`to_datetime` with ``arg:xr.DataArray`` and ``unit="ns"`` specified raises TypeError (:issue:`44053`)
 - Bug in :meth:`DataFrame.convert_dtypes` not returning the correct type when a subclass does not overload :meth:`_constructor_sliced` (:issue:`43201`)
--
+- Bug in :meth:`NDFrame.astype` not propagating ``attrs`` from the original :class:`DataFrame` (:issue:`44414`)
```
Comment (phofl) [inline doc/source/whatsnew/v1.4.0.rst:639]: Please use DataFrame not NDFrame

File: pandas/core/generic.py [hunk 2]
```diff
@@ -5890,6 +5890,8 @@ def astype(
         # GH 19920: retain column metadata after concat
         result = concat(results, axis=1, copy=False)
         result.columns = self.columns
+        # GH#44414: call finalize to retain metadata for the DataFrame or Series
```
Comment (phofl) [inline pandas/core/generic.py:5893]: Can remove the comment

File: pandas/tests/frame/methods/test_astype.py [hunk 3]
```diff
@@ -714,6 +714,15 @@ def test_astype_noncontiguous(self, index_slice):
         expected = df.iloc[index_slice]
         tm.assert_frame_equal(result, expected, check_dtype=False)
 
+    def test_astype_retain_attrs(self):
+        # GH#44414
+        df = DataFrame({"a": [0, 1, 2], "b": [3, 4, 5]})
+        df.attrs["Location"] = "Michigan"
+
+        df2 = df.astype({"a": float})
```
Comment (phofl) [inline pandas/tests/frame/methods/test_astype.py:722]: Please call result

File: pandas/tests/frame/methods/test_astype.py [hunk 4]
```diff
@@ -714,6 +714,15 @@ def test_astype_noncontiguous(self, index_slice):
         expected = df.iloc[index_slice]
         tm.assert_frame_equal(result, expected, check_dtype=False)
 
+    def test_astype_retain_attrs(self):
```
Comment (phofl) [inline pandas/tests/frame/methods/test_astype.py:717]: Could you check if we have tests for the other cases. If not, please parametrize here

**Review by phofl — CHANGES_REQUESTED** [review]

**Comment by phofl** [comment]
Couple of comments

File: doc/source/whatsnew/v1.4.0.rst (hunk 1 above)
Comment (skwirskj) [reply doc/source/whatsnew/v1.4.0.rst:639]: Got it!

File: pandas/core/generic.py (hunk 2 above)
Comment (skwirskj) [reply pandas/core/generic.py:5893]: removed

File: pandas/tests/frame/methods/test_astype.py (hunk 4 above)
Comment (skwirskj) [reply pandas/tests/frame/methods/test_astype.py:717]: Just to clarify, we would want a case for changing all columns' data type correct?

Comment (phofl) [reply pandas/tests/frame/methods/test_astype.py:717]: No, there are different options of parametrizing astype, a data type, data type represented as string or a dict of one of both

Comment (skwirskj) [reply pandas/tests/frame/methods/test_astype.py:717]: Oh ok, I see, I will double-check for this and add any tests if necessary.

Comment (skwirskj) [reply pandas/tests/frame/methods/test_astype.py]: Added the parametrization

File: doc/source/whatsnew/v1.4.0.rst [hunk 5]
```diff
@@ -636,7 +636,7 @@ Conversion
 - Bug in :class:`IntegerDtype` not allowing coercion from string dtype (:issue:`25472`)
 - Bug in :func:`to_datetime` with ``arg:xr.DataArray`` and ``unit="ns"`` specified raises TypeError (:issue:`44053`)
 - Bug in :meth:`DataFrame.convert_dtypes` not returning the correct type when a subclass does not overload :meth:`_constructor_sliced` (:issue:`43201`)
--
+- Bug in :meth:`DataFrame.astype` not propagating ``attrs`` from the original :class:`DataFrame` (:issue:`44414`)
```
Comment (jreback) [inline doc/source/whatsnew/v1.4.0.rst:639]: can you move this next to the other notes about attr changes

**Review by jreback — CHANGES_REQUESTED** [review]
looks fine, small comment. we have a comprehensive test for methods, odd that this doesn't need changing (as it would not xpass)

File: doc/source/whatsnew/v1.4.0.rst (hunk 5 above)
Comment (skwirskj) [reply doc/source/whatsnew/v1.4.0.rst:639]: Yes I can, I only found mentions of Attributes being changed within I/O and sparse section. Should I move it to one of those? Thank you in advance for the help

File: pandas/tests/frame/methods/test_astype.py [hunk 6]
```diff
@@ -714,6 +714,19 @@ def test_astype_noncontiguous(self, index_slice):
         expected = df.iloc[index_slice]
         tm.assert_frame_equal(result, expected, check_dtype=False)
 
+    @pytest.mark.parametrize(
+        "dtype", ["float", "int32", "int64", "string", "int16", "float32"]
```
Comment (jreback) [inline pandas/tests/frame/methods/test_astype.py:718]: can you use one of the fixtures in pandas/conftest.py instead

**Review by jreback — CHANGES_REQUESTED** [review]
add a test for series as well. merge master and ping on green.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @skwirskj

**Comment by skwirskj** [comment]
No problem @jreback thank you for your help!

**Pull request merged**